wunderio/next-drupal-starterkit
# NEX-155: ddev support

This PR adds support for ddev for the starterkit.

### Changes introduced

* There are now two setup scripts, one for lando, one for ddev
* I replicated our custom lando commands (to install a recipe, to set up oauth keys) for ddev

For the ddev setup, I used this article as a guide, and I went with the first option described there: https://www.velir.com/ideas/2024/05/13/how-to-run-headless-drupal-and-nextjs-on-ddev?utm_source=linkedin&utm_medium=social&sc_camp=FFCB08E97FEE4B629943BBC8546739F4

So we are not setting up a different container, but using the web one on a different port. Also, I didn't set up a daemon like they suggest in the article, because I would like this to work like it does in lando, where to see the site you have to use either `ddev npm run dev` or `ddev npm run start`, since this is used to test different things.

One difference is that for the `ddev npm` command to work, you need to cd to the `next` folder first. I tried to highlight this in the README file, but I can see this confusing people.

### How to test

1. [install ddev if you don't have it](https://ddev.readthedocs.io/en/stable/users/install/ddev-installation/)
2. run `./setup-ddev.sh`
3. open backend, see that preview works
4. open frontend (see provided url, note that localhost:3000 does not work
5. open elasticvue, add predefined cluster check that you can see elasticsearch contents
6. `ddev describe` should say this: 
```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Project: next-drupal-starterkit ~/www/next-drupal-starterkit https://next-drupal-starterkit.ddev.site                                                                         │
│ Docker platform: docker-desktop                                                                                                                                               │
│ Router: traefik                                                                                                                                                               │
├────────────┬──────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬──────────────────────┤
│ SERVICE    │ STAT │ URL/PORT                                                                                                                           │ INFO                 │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ web        │ OK   │ https://next-drupal-starterkit.ddev.site                                                                                           │ drupal PHP8.2        │
│            │      │ InDocker: web:3000,443,80,8025                                                                                                     │ nginx-fpm            │
│            │      │ Host: 127.0.0.1:61141,61140                                                                                                        │ docroot:'drupal/web' │
│            │      │                                                                                                                                    │ Perf mode: mutagen   │
│            │      │                                                                                                                                    │ NodeJS:20            │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ db         │ OK   │ InDocker: db:3306                                                                                                                  │ mariadb:10.11        │
│            │      │ Host: 127.0.0.1:61139                                                                                                              │ User/Pass: 'db/db'   │
│            │      │                                                                                                                                    │ or 'root/root'       │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ elasticsea │ OK   │ https://next-drupal-starterkit.ddev.site:9201                                                                                      │                      │
│ rch        │      │ InDocker: elasticsearch:9200,9200,9300,9300                                                                                        │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ elasticvue │ OK   │ http://next-drupal-starterkit.ddev.site:9003                                                                                       │                      │
│            │      │ InDocker: elasticvue:80                                                                                                            │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ Mailpit    │      │ Mailpit: https://next-drupal-starterkit.ddev.site:8026                                                                             │                      │
│            │      │ `ddev mailpit`                                                                                                                     │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ Next.js    │      │ InDocker: localhost:3000 https://next-drupal-starterkit.ddev.site:3000 http://next-drupal-starterkit.ddev.site:2999                │                      │
├────────────┼──────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼──────────────────────┤
│ All URLs   │      │ https://next-drupal-starterkit.ddev.site, https://127.0.0.1:61140, http://next-drupal-starterkit.ddev.site, http://127.0.0.1:61141 │                      │
└────────────┴──────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴──────────────────────┘

 ```

### Open questions

Something that is problematic I think, is that `npm run dev` will print localhost:3000 as the url that the frontend runs in, but that path is not accessible. you have to use `https://next-drupal-starterkit.ddev.site:3000` instead, and that information is not so easily accessible...

I wonder if we can improve this somehow?
